### PR TITLE
React to aspnet/Razor#279 changes.

### PR DESCRIPTION
**Pull request by NTaylorMullen**
- Updated all tests and TagHelpers to utilize the new TagHelperOutput.Attributes and TagHelperContext.AllAttribute types.

aspnet/Razor#279

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs [hunk 1]
```diff
@@ -31,21 +31,20 @@ public static class TagHelperOutputExtensions
...
+                if (!context.AllAttributes.TryGetAttribute(attributeName, out entry))
                 {
                     throw new ArgumentException(
                         Resources.FormatTagHelperOutput_AttributeDoesNotExist(attributeName, nameof(TagHelperContext)),
                         nameof(attributeName));
                 }
-                
-                tagHelperOutput.Attributes.Add(entry.Key, entry.Value);
+
+                tagHelperOutput.Attributes.Add(entry.Name, entry.Value);
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs]: this method should restore all attributes with given name.  aspnet/Razor#279 is incomplete as long as this method only handles the first match.

File: src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs [hunk 2]
```diff
@@ -85,13 +84,22 @@ public static class TagHelperOutputExtensions
...
                     tagHelperOutput.Attributes.Add(attribute.Key, attribute.Value);
                 }
                 else if (attribute.Key.Equals("class", StringComparison.OrdinalIgnoreCase))
                 {
-                    tagHelperOutput.Attributes["class"] += " " + attribute.Value;
+                    TagHelperAttribute classAttribute;
+
+                    if (tagHelperOutput.Attributes.TryGetAttribute("class", out classAttribute))
+                    {
+                        classAttribute.Value += " " + attribute.Value;
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs]: use `tagHelperOutput.Attributes["class"] = classAttribute.Value + " " + attribute.Value;` to remove other attributes named `class`

File: src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs [hunk 3]
```diff
@@ -85,13 +84,22 @@ public static class TagHelperOutputExtensions
...
+                    }
+                    else
+                    {
+                        tagHelperOutput.Attributes["class"] = attribute.Value;
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs]: be consistent for the addition case: `tagHelperOutput.Attributes.Add("class", attribute.Value)`

File: test/Microsoft.AspNet.Mvc.Razor.Test/RazorPageTest.cs [hunk 4]
```diff
@@ -782,7 +782,7 @@ public void IsSectionDefined_ThrowsIfPreviousSectionWritersIsNotRegistered()
                     {
                         GetTagHelperOutput(
                             tagName:     "p",
-                            attributes:  new Dictionary<string, object>() { { "test", "testVal" } },
+                            attributes:  new TagHelperAttributes() { ["test"] = "testVal" },
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Razor.Test/RazorPageTest.cs]: revert unnecessary churn.  you've got an `Add(string, object)` method available; no reason to switch to the long-form init syntax.  if you prefer this syntax and others agree, then do as a separate, broader PR.

File: test/Microsoft.AspNet.Mvc.TagHelpers.Test/LinkTagHelperTest.cs [hunk 5]
```diff
@@ -202,21 +202,21 @@ public void PreservesOrderOfSourceAttributesWhenRun()
...
-                    ["asp-fallback-test-class"] = "hidden",
-                    ["asp-fallback-test-property"] = "visibility",
-                    ["asp-fallback-test-value"] = "hidden"
+                    { "rel", new HtmlString("stylesheet") },
+                    { "data-extra", new HtmlString("something") },
+                    { "href", "test.css" },
+                    { "asp-fallback-href", "test.css" },
+                    { "asp-fallback-test-class", "hidden" },
+                    { "asp-fallback-test-property", "visibility" },
+                    { "asp-fallback-test-value", "hidden" },
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.TagHelpers.Test/LinkTagHelperTest.cs]: revert unnecessary churn.  with the exception of the `HtmlString` cases, stick with the current long-form init syntax.

**Comment by dougbu** [comment]
needs additional test cases involving attributes in the Razor source with duplicate names: identical names and just case-insensitive matches in both unit tests (especially involving tag helpers that recreate elements e.g. `<link/>` or restore attributes e.g. `<form/>`) and functional tests. otherwise are not confirming MVC uses Razor runtime correctly.

**Comment by dougbu** [comment]
:watch:

File: src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs (hunk 1 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.TagHelpers/TagHelperOutputExtensions.cs]: :+1:

File: test/Microsoft.AspNet.Mvc.TagHelpers.Test/LinkTagHelperTest.cs (hunk 5 above)
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Mvc.TagHelpers.Test/LinkTagHelperTest.cs]: Invalid initializer member declarator when intermingling the two. Not going to look into why. Just going to leave for now, no harm.

**Comment by NTaylorMullen** [comment]
Updated.

**Comment by dougbu** [comment]
:shipit: once aspnet/Razor#279 is in

File: src/Microsoft.AspNet.Mvc.TagHelpers/InputTagHelper.cs [hunk 6]
```diff
@@ -218,9 +218,16 @@ private void GenerateCheckBox(ModelExplorer modelExplorer, TagHelperOutput outpu
             }
 
             // Prepare to move attributes from current element to <input type="checkbox"/> generated just below.
-            var htmlAttributes = output.Attributes.ToDictionary(
-                attribute => attribute.Key,
-                attribute => (object)attribute.Value);
+            var htmlAttributes = new Dictionary<string, object>(StringComparer.OrdinalIgnoreCase);
+
+            // Construct attributes correctly (first attribute wins).
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.TagHelpers/InputTagHelper.cs]: oops, this is where the :dancers: danced
